### ISSUE TITLE
Coroutine implementation throws uncaught exception instead of FuelError

### ISSUE THREAD

**Issue by markGilchrist** [kittinunf/Fuel#358]
**Desired**

As a user of the coroutine library I would like to see any server error, or deserialisation error  returned to me as part of the result object ( `Result<T,FuelError>`).

**Actual**

When a server error is thrown then an uncaught exception is thrown and my app crashes
When a deserialisation error is thrown then an uncaught exception is thrown and my app crashes

I believe the the line of code causing this is Coroutines.kt, line 27

I have removed the source code from my gradle and created a local file in which I have tested the following modifications, if you feel it is an improvement I will happily make a pr with the appropriate
Unit tests and submit it for review.

```
private suspend fun <T : Any, U : Deserializable<T>> Request.awaitSafely(
        deserializable: U
): Triple<Request, Response, Result<T, FuelError>> =
        suspendCancellableCoroutine { continuation ->
            continuation.invokeOnCompletion {
                if (continuation.isCancelled) {
                    continuation.cancel()
                }
            }
            continuation.resume(response(deserializable))
        }
...


/**
 * This function catches both server errors and Deserialization Errors
 *
 * @param deserializable
 *
 * @return Result object
 * */
suspend fun <U : Any> Request.awaitResultObject (
        deserializable: ResponseDeserializable<U>
):  Result<U, FuelError> = awaitSafely(deserializable).third

```

If this is not a suitable solution then could we mark this as a `@Throws`?

If this is not an issue then please explain what I have done incorrectly. 
thanks mark

**Comment by D-A-R** [comment]
Just ran into the same issue. 

Changing from the synchronous 
get().response(): Triple<request, response, result>
to
get().awaitResponse(): Triple<request, response, result>

We started to see a difference in behaviour in the app, which traced back to this issue that no longer were we getting a result so to speak, but it was now throwing an exception

We tried changing the signature of await too to:
```
private suspend fun <T : Any, U : Deserializable<T>> Request.await(
        deserializable: U
): Triple<Request, Response, Result<T, FuelError>> =
        suspendCancellableCoroutine { continuation ->
            continuation.invokeOnCompletion {
                if (continuation.isCancelled) {
                    cancel()
                }
            }

            response(deserializable) { request: Request, response: Response, result: Result<T, FuelError> ->
                continuation.resume(Triple(request, response, result))
            }
        }
```
to be consistent with the synchronous implementation whereby a result is always returned. 

If you do this, the next hurdle you run into is the implementation of Result<>. If the result is an error and you perform a get() on that result it throws the error, which is what we're actually seeing. As to whether this is correct behaviour for a Result<> is another question.

Now, each of the async Result based methods, `awaitResponseResult`, `awaitStringResult` and `awaitObjectResult` all directly call get() on .third() which then throws an exception.

The first 2 can easily be changed to their defaults, empty byte array and empty string, but object is an interesting one?

**Comment by markGilchrist** [comment]
@D-A-R you sound like you have got a bit further than I have, If you look at the unit tests in `coroutinesTest.kt`, line 88 you can see that the @lucasvalenteds has written the test, expecting an exception to be throw. 

I have looked at the deserialiseable.kt 

would something like be a better approach?

I haven't tried it and I would involve enabling coroutines in the main module which the authors may be against

```
suspend fun <T : Any, U : Deserializable<T>> Request.awaitResponse(deserializable: U): Triple<Request, Response, Result<T, FuelError>> {
   return  suspendCancellableCoroutine { continuation ->
        continuation.invokeOnCompletion {
            if (continuation.isCancelled) {
                cancel()
            }
        }
        try {
            val response = taskRequest.call()
            continuation.resume(Triple(this, response, Result.Success(deserializable.deserialize(response))))
        } catch (error: FuelError) {
            continuation.resume(Triple(this, error.response, Result.error(error)))
        }
    }
} 
```

**Comment by lucasvalenteds** [comment]
Hi @markGilchrist. Thank you for raising the issue. I'm sorry for the late response, I just received the notification after um mentioned me.

So about the problem, have you tried to catch an exception of the serialization/deserialization library? 

For example, the test bellow implements a deserializer using Jackson. Since the endpoint `/uuid` expects a class with property `uuid` and the class `NotUUID` doesn't have it, the deserialization exception can be handled by `JsonMappingException`.

```kotlin
private data class NotUUID(val notUuid: Int)

private object NotUUIDDeserializer : ResponseDeserializable<NotUUID> {
    override fun deserialize(content: String) =
            jacksonObjectMapper().readValue<NotUUID>(content)
}

@Test
fun testDeserializationExceptionCanBeHandled() = runBlocking<Unit> {
    try {
        Fuel.get("/uuid").awaitObjectResult(NotUUIDDeserializer)
        fail("This test should fail due to missing `uuid` property of class NotUUID.")
    } catch (ex: JsonMappingException) {
        assertNotNull(ex)
    }
}
```

**Comment by markGilchrist** [comment]
@lucasvalenteds very quick response as I only tagged you in yesterday and thank you for taking the time. 

I haven't yet tried changing continuation.resumeWithException(it.exception) to continuation.resumeWithException(it) but I will

My question was also would it not be better to have deserialisation errors caught in this manner out of the box? I was expecting that behaviour, however I would say I am a new 'lover' and user of fuel so my expectations are may not be in line with the approach of this awesome lib.

If there is any why I can make any contributions then please say, keep up the great work

**Comment by lucasvalenteds** [comment]
> My question was also would it not be better to have deserialisation errors caught in this manner out of the box?

Could you please check the issue #365 and see if it's what you're suggesting? At this time I'm waiting response of another user about how to handle exceptions.

> If there is any why I can make any contributions then please say

Since it involves design/consistency of the library some maintainer could help with that. @kittinunf Could you take a look, please?

**Comment by lucasvalenteds** [comment]
@markGilchrist It's been a while since the last comment. Do you think you can send a PR with the fix?

**Comment by markGilchrist** [comment]
@lucasvalenteds I will do it asap, I have some questions can you email me mark[dot]s[dot]w[dot]gilchrist[at]gmail[dot]com

**Comment by raharrison** [comment]
Sorry to jump on this issue, but I just opened issue #368 which is about some API changes in the newer version of Kotlin coroutines. If you planning on making a PR for this, it might be good to upgrade and remove the deprecated call at the same time. I'm unsure if this might have some side effects for these fixes so it might save us from doing the work twice.

**Comment by markGilchrist** [comment]
@raharrison  please contact me on slack or email mark[dot]s[dot]w[dot]gilchrist[at]gmail[dot]com

I have started to make some changes and create a pr 

https://github.com/kittinunf/Fuel/compare/master...markGilchrist:pass_all_exceptions_to_as_Fuel_error_coroutines 

comments and suggestions would be warmly received

**Comment by markGilchrist** [comment]
this issue is closed

**Comment by mobiletoly** [comment]
So why this issue was closed? I'm trying 1.16.0 with .awaitObject() and it still throws HttpException and I don't have access to "httpCode" field that I need.

**Comment by SleeplessByte** [comment]
It's fixed in master.
